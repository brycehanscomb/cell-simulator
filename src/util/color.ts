/**
 * Change the brightness of a six-digit hex color.
 *
 * Stolen from David Sherret via:
 * @see https://stackoverflow.com/a/13532993
 *
 * @param color - a hex color in the format of `#RRGGBB`
 * @param percent - an int from -100 to 100
 */
export function shadeColor(color: string, percent: number) {
  let R: number | string = parseInt(color.substring(1, 3), 16);
  let G: number | string = parseInt(color.substring(3, 5), 16);
  let B: number | string = parseInt(color.substring(5, 7), 16);

  R = parseInt(String((R * (100 + percent)) / 100));
  G = parseInt(String((G * (100 + percent)) / 100));
  B = parseInt(String((B * (100 + percent)) / 100));

  R = R < 255 ? R : 255;
  G = G < 255 ? G : 255;
  B = B < 255 ? B : 255;

  R = R.toString(16);
  G = G.toString(16);
  B = B.toString(16);

  const RR = R.length === 1 ? "0" + R : R;
  const GG = G.length === 1 ? "0" + G : G;
  const BB = B.length === 1 ? "0" + B : B;

  return `#${RR}${GG}${BB}`;
}
