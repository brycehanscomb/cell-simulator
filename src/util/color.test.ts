import { shadeColor } from "./color";

describe("Utils::color", () => {
  describe("shadeColor", () => {
    it("should return black when going 100% darker", () => {
      const inputs = ["#05A6B7", "#FF0000", "#FFFFFF"];

      inputs.forEach(input => {
        expect(shadeColor(input, -100)).toEqual("#000000");
      });
    });

    it("should correctly darken images by the specified amount when `percent` is negative", () => {
      const inputs = ["#05A6B7", "#FF0000", "#FFFFFF"];

      /**
       * We've verified the values in the snapshots match the correct outputs.
       */
      inputs.forEach(input => {
        expect(shadeColor(input, -25)).toMatchSnapshot();
      });
    });

    it("should correctly lighten images by the specified amount when `percent` is positive", () => {
      const inputs = ["#05A6B7", "#FF0000", "#00FF25"];

      /**
       * We've verified the values in the snapshots match the correct outputs.
       */
      inputs.forEach(input => {
        expect(shadeColor(input, 25)).toMatchSnapshot();
      });
    });

    /**
     * Other valuable tests that could be done here could include:
     *
     * - What if the input is empty?
     * - What if input is a 3-digit hex code instead of 6-digit?
     * - What if the input is an 8-digit hex code with an alpha channel?
     * - Are the hex letters case-sensitive?
     * - What if we pass zero as the change amount?
     * - What if we exceed the -100 to 100 percentage bounds?
     */
  });
});
