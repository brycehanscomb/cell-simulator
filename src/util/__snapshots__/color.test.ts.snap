// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utils::color shadeColor should correctly darken images by the specified amount when \`percent\` is negative 1`] = `"#037c89"`;

exports[`Utils::color shadeColor should correctly darken images by the specified amount when \`percent\` is negative 2`] = `"#bf0000"`;

exports[`Utils::color shadeColor should correctly darken images by the specified amount when \`percent\` is negative 3`] = `"#bfbfbf"`;

exports[`Utils::color shadeColor should correctly lighten images by the specified amount when \`percent\` is positive 1`] = `"#06cfe4"`;

exports[`Utils::color shadeColor should correctly lighten images by the specified amount when \`percent\` is positive 2`] = `"#ff0000"`;

exports[`Utils::color shadeColor should correctly lighten images by the specified amount when \`percent\` is positive 3`] = `"#00ff2e"`;
